debugger_implementation.tex

```diff
@@ -615,3 +615,32 @@ \subsection{Single Step} \label{nativestep}
     lw    sp, 0(sp)     /* Restore sp, count decrements to 2 */
     mret                /* Return to program being debugged. count decrements to 1 */
 \end{verbatim}
+
+There is an additional problem with using \RcsrIcount to single step.  An
+instruction may cause an exception into a more privileged mode where the trigger
+is not enabled.  The exception handler might address the cause of the exception,
+and then restart the instruction.  Examples of this include page faults and FPU
+instructions when the FPU is not yet enabled.  When a user is single stepping
+through such code, they will have to step twice to get past the restarted
+instruction. The first time the exception handler runs, and the second time the
+instruction actually executes. That is confusing and usually undesirable.
+
+To help users out, debuggers should detect when a single step restarted an
+instruction, and then step again. This way the users see the expected behavior
+of stepping over the instruction. Ideally the debugger would notify the user
+that an exception handler executed the first time.
+
+\begin{steps}{The debugger should perform this extra step if all the following
+        conditions are satisfied:}
+    \item The PC did not change during the first step.
+    \item The PC does not equal the {\tt epc} register for the current privilege mode.
```

Comment:
This condition isn't immediately obvious to me.  If you're stepping S mode, PC=X, you step, and PC is still X then step one more time.  Why do we also need to check if sepc=X?

Reply:
That should be `tvec`, not `epc`.

The condition I'm worried about is if you're stepping the first instruction in an exception handler, that itself causes that exception handler to be called. But maybe that doesn't matter similar to jal to self makes no changes after the first invocation.

I'd love some more input into my thinking here.

Reply:
tvec makes more sense.  Say that you're stepping U+S modes and breakpoints go to M.  The following happens:
- You step X and it traps.
- The trap sends you to M mode with mepc=X.
- PC did change so you wait for the debugger user to request another step.  (Don't automatically do the extra step you're describing.)
- User says to step.  Your vector is an illegal instruction (or whatever) so it traps.
- The trap sends you to M mode with mepc=stvec.

You're saying that the debugger should not do the extra step here.  I agree that you don't want to do the extra step.  However, if you did then you'd hit the same illegal instruction and come back to M mode with mepc=stvec so there's no real harm.